gobblin-service/src/main/java/org/apache/gobblin/service/modules/db/ServiceDatabaseProviderImpl.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gobblin.service.modules.db;
+
+import java.time.Duration;
+import java.util.Objects;
+
+import org.apache.commons.dbcp2.BasicDataSource;
+
+import com.typesafe.config.Config;
+
+import javax.inject.Inject;
+import javax.sql.DataSource;
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+
+import org.apache.gobblin.password.PasswordManager;
+import org.apache.gobblin.service.ServiceConfigKeys;
+import org.apache.gobblin.util.ConfigUtils;
+
+
+public class ServiceDatabaseProviderImpl implements ServiceDatabaseProvider {
+
+  private final Configuration configuration;
+  private BasicDataSource dataSource;
+
+  @Inject
+  public ServiceDatabaseProviderImpl(Configuration configuration) {
+    this.configuration = configuration;
+  }
+
+  public DataSource getDatasource() {
+    ensureDataSource();
+    return dataSource;
+  }
+
+  private synchronized void ensureDataSource() {
```

Comment:
For all other usages of mysql in gobblin and gaas code, we use `MysqlDataSourceFactory` (which calls `MysqlStateStore.newDataSource`) to create the datasource. Is the configuration for this one unique enough that it needs to be a separate thing, or that these configurations would not also be useful in that place?

If we do want to keep these separate, maybe we could at least reuse the validation query (which I just added to `MysqlStateStore.newDataSource`) so that we don't have the same query in two places.

Reply:
I used it initially, but there are problems with configuring it. It expects settings like "state.store.db.user" and "state.store.db.table" which don't really make sense for the issues and other tables. What we have in service db is not limited to one table, and it's not a state store.

I'll put the query in utility class to be reused.

Reply:
Just noting, for other GaaS tables we separate them by using a config prefix like `gobblin.service.mysqlDagStateStore.state.store.db.table`. And we do also use `state.store.db.` prefixed configs even for the `MySqlSpecStore`, which is not an implementation of a state store. But yeah that's not really the best pattern, so I think it's fine to have this one separate. If anything maybe we should change the spec store to use this datasource instead at some point.